alf/networks/preprocessor_networks.py

```diff
@@ -72,30 +72,34 @@ def _get_preprocessed_spec(preproc, spec):
                 # In this case we just assume the spec won't change after the
                 # preprocessing. If it does change, then you should consider
                 # defining an input preprocessor network instead.
-                assert not isinstance(
-                    preproc, nn.Module), ("To use trainable preprocessors, "
-                                          "please derive from alf.Network")
                 return spec
             return preproc.output_spec
 
         self._input_preprocessors = None
         if input_preprocessors is not None:
-            input_tensor_spec = alf.nest.map_structure(
-                _get_preprocessed_spec, input_preprocessors, input_tensor_spec)
 
-            def _return_or_copy_preprocessor(preproc):
+            def _return_or_copy_preprocessor(preproc, input_spec):
                 if preproc is None:
                     # allow None as a placeholder in the nest
-                    return math_ops.identity
+                    return lambda x: (x, None)
                 elif isinstance(preproc, Network):
                     preproc = preproc.copy()
                     self._input_preprocessor_modules.append(preproc)
-                    return lambda x: preproc(x)[0]
-                else:
                     return preproc
+                elif isinstance(preproc, nn.Module):
+                    preproc = SequentialNetwork(input_spec, [preproc])
```

Comment:
In this case, if we call ```preprocesso_network.copy()```, will/should ```preproc``` be copied?

Reply:
preproc will be copied, same as Network as input processor.

Reply:
The current implementation will not copy preproc in the case as follows
```
        preproc = torch.nn.Linear(10, 10)

        net = EncodingNetwork(input_tensor_spec=TensorSpec((10, )),
                              input_preprocessors=preproc,
                              fc_layer_params=(10, 10))

        copied_net = net.copy()
```

```preproc``` will be shared between ```net``` and ```copied_net```.

Reply:
You are right. What behavior do you think is reasonable? Or we can remove the ability of using nn.Module here. The user can always use SequentialNetwork himself in the gin.

Reply:
I think we should make the expected behavior consistent with others, i.e., should be copied in this case as you said.
Being able to use nn.Module would be good to make the gin file simpler.
Maybe we can do ```preproc = SequentialNetwork(input_spec, [preproc]).copy()``` for now?